nvflare/edge/assessors/buff_device_manager.py

```diff
@@ -36,18 +40,121 @@ def __init__(
             device_reuse (bool): Whether to allow reusing devices across different model versions. Defaults to True.
+            device_sampling_strategy (str): Strategy for sampling devices when filling selection. Defaults to "balanced".
+                - "balanced": try to balance the usage of devices across clients.
+                - "random": randomly select devices from the available pool.
         """
         super().__init__()
         check_positive_int("device_selection_size", device_selection_size)
         check_positive_int("min_hole_to_fill", min_hole_to_fill)
+        check_positive_int("initial_min_client_num", initial_min_client_num)
 
         self.device_selection_size = device_selection_size
+        self.initial_min_client_num = initial_min_client_num
         self.min_hole_to_fill = min_hole_to_fill
         self.device_reuse = device_reuse
+        self.device_sampling_strategy = device_sampling_strategy
         # also keep track of the current selection version and used devices
         self.current_selection_version = 0
+        self.used_devices = {}
+        # keep a map of device_id -> client_name
+        self.device_client_map = {}
+
+    def _balanced_device_sampling(self, usable_devices: Set[str], num_holes: int) -> Set[str]:
+        """Sample devices while balancing across clients.
+
+        Args:
+            usable_devices: Set of device IDs that can be selected
+            num_holes: Number of devices to sample
+
+        Returns:
+            Set of selected device IDs
+        """
+        if not usable_devices or num_holes <= 0:
+            return set()
+
+        # Count devices per client efficiently using Counter
+        client_device_counts = Counter(
+            self.device_client_map[device_id] for device_id in usable_devices if device_id in self.device_client_map
+        )
+
+        # Group devices by client using defaultdict for efficiency
+        client_devices = defaultdict(list)
+        for device_id in usable_devices:
+            if device_id in self.device_client_map:
+                client_devices[self.device_client_map[device_id]].append(device_id)
+
+        if not client_device_counts:
+            # Fallback to random sampling if no client mapping
+            return set(random.sample(list(usable_devices), min(num_holes, len(usable_devices))))
+
+        # Randomize client order for more balanced distribution
+        clients_list = list(client_device_counts.items())
+        random.shuffle(clients_list)
+
+        selected_devices = set()
+        remaining_holes = num_holes
+
+        # First pass: assign minimum possible to each client
+        min_per_client = remaining_holes // len(clients_list)
+        extra_holes = remaining_holes % len(clients_list)
+
+        for i, (client_name, device_count) in enumerate(clients_list):
+            # Calculate how many devices this client should get
```

Comment:
so we mean "how many devices this client should use (regardless of how many devices connected to this client)"?

Reply:
no it cannot exceed what a clint have